### PR TITLE
support create space as

### PR DESCRIPTION
**Pull request by liuyu85cn**
like create table as ... 

create a new space as an existing one.

create space <new_space_name> as <old_space_name>, tag, edge and index will also be created. 

this will generate meta only.
```
(root@nebula) [(none)]> create space ldbc2 as ldbc_snb_sf10
Execution succeeded (time spent 11777/12414 us)

Tue, 31 Aug 2021 15:03:59 CST
```

### CODE REVIEW

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    return;
+  }
+
+  if (nebula::ok(newSpaceId)) {
+    if (!req.get_if_not_exists()) {
+      LOG(ERROR) << "Create Space " << newSpaceName << " as " << oldSpaceName << "failed. "
+                 << newSpaceName << " already exist";
+      rc_ = nebula::cpp2::ErrorCode::E_EXISTED;
+      return;
+    }
```
Comment (panda-sheep) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:41]: When` req.get_if_not_exists()` is true, return success directly？
But this new space is not the desired space anymore.
Maybe `if_not_exists `is not needed?

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    return nebula::error(tagPrefix);
+  }
+
+  std::vector<kvstore::KV> data;
+  auto iter = nebula::value(tagPrefix).get();
+  for (; iter->valid(); iter->next()) {
+    auto val = iter->val();
+
+    auto tagId = autoIncrementIdInSpace(newSpaceId);
```
Comment (panda-sheep) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:138]: Is the tagId in the new space different for the same tag in the old space?

**Review by panda-sheep — COMMENTED** [review]
Good job

`create table A like B in mysql`
Only copy the table structure of B.
`create table A as..`
Not only copy the table structure, but also copy the data.

maybe `like `is more appropriate

**Review by yixinglu — COMMENTED** [review]
Please add TCK test cases

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp (hunk 1 above)
Comment (liuyu85cn) [reply src/meta/processors/parts/CreateSpaceAsProcessor.cpp]: good catch, removed.

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    onFinished();
+  };
+
+  folly::SharedMutex::WriteHolder wHolder(LockUtils::spaceLock());
+  auto oldSpaceName = req.get_old_space_name();
+  auto newSpaceName = req.get_new_space_name();
+  auto oldSpaceId = getSpaceId(oldSpaceName);
+  auto newSpaceId = getSpaceId(newSpaceName);
+
```
Comment (panda-sheep) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:27]: When the newspace exists, directly report the existing error？

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  data.emplace_back(MetaServiceUtils::indexSpaceKey(spaceName),
+                    std::string(reinterpret_cast<const char *>(&newSpaceId), sizeof(newSpaceId)));
+  cpp2::SpaceDesc spaceDesc = MetaServiceUtils::parseSpace(nebula::value(oldSpaceVal));
+  spaceDesc.set_space_name(spaceName);
+  auto *typeLenPtr = spaceDesc.get_vid_type().get_type_length();
+  if (typeLenPtr) {
+    vIdLen_ = *typeLenPtr;
+  }
+  data.emplace_back(MetaServiceUtils::spaceKey(newSpaceId), MetaServiceUtils::spaceVal(spaceDesc));
+  return data;
```
Comment (panda-sheep) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:126]: Good！

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+  const std::string kIdKey = MetaServiceUtils::idKey();
+  std::string val;
+  rc_ = kvstore_->get(kDefaultSpaceId, kDefaultPartId, kIdKey, &val);
+  if (rc_ == nebula::cpp2::ErrorCode::SUCCEEDED) {
+    int32_t id = *reinterpret_cast<const int32_t *>(val.c_str());
+    data.emplace_back(kIdKey, std::string(reinterpret_cast<const char *>(&id), sizeof(id)));
+  } else {
+    return;
+  }
```
Comment (critical27) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:89]: Why we write `kIdKey` again? I remember `autoIncrementId` has already write.

**Review by critical27 — COMMENTED** [review]
Good job, generally LGTM, add a simple tck case.

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+  rc_ = LastUpdateTimeMan::update(kvstore_, time::WallClock::fastNowInMilliSec());
+  if (rc_ != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    LOG(ERROR) << "update last update time error, " << apache::thrift::util::enumNameSafe(rc_);
+    return;
+  }
+  LOG(INFO) << "created space " << newSpaceName;
+}
```
Comment (bright-starry-sky) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:104]: what about `doSyncPutAndUpdate` ?

File: tests/tck/conftest.py [hunk 7]
```diff
@@ -203,6 +202,17 @@ def try_to_execute_query(query, graph_spaces, session, request):
     for stmt in ngql.split(';'):
         exec_query(request, stmt, session, graph_spaces, True)
 
+@when(parse("clone a new space according to current space:\n{new_space}"))
```
Comment (yixinglu) [inline tests/tck/conftest.py:205]: I don’t think there needs a new_space parameter here. Random generate the space name.

Comment (critical27) [reply tests/tck/conftest.py:205]: 👌

**Review by yixinglu — DISMISSED** [review]
LGTM

**Review by critical27 — DISMISSED** [review]

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+      handleErrorCode(rc_);
+    }
+    onFinished();
+  };
+
+  folly::SharedMutex::WriteHolder wHolder(LockUtils::spaceLock());
+  auto oldSpaceName = req.get_old_space_name();
+  auto newSpaceName = req.get_new_space_name();
+  auto oldSpaceId = getSpaceId(oldSpaceName);
+  auto newSpaceId = getSpaceId(newSpaceName);
```
Comment (bright-starry-sky) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:26]: `if (nebula::ok(newSpaceId)) {
    auto ret = nebula::cpp2::ErrorCode::SUCCEEDED;
    if (!req.get_if_not_exists()) {
      LOG(ERROR) ...
      ret = nebula::cpp2::ErrorCode::E_EXISTED;
    }
    resp_.set_id(...);
    handleErrorCode(ret);
    onFinished();
    return;
  }`

Comment (critical27) [reply src/meta/processors/parts/CreateSpaceAsProcessor.cpp:26]: Maybe we should not `support if_not_exists` in this sentence?

Comment (panda-sheep) [reply src/meta/processors/parts/CreateSpaceAsProcessor.cpp:26]: When newSpaceName exists, an error should be reported directly

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline src/meta/processors/parts/CreateSpaceAsProcessor.cpp:1]: 2021

File: src/meta/processors/parts/CreateSpaceAsProcessor.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (darionyaphet) [inline src/meta/processors/parts/CreateSpaceAsProcessor.h:1]: ditto

File: src/meta/processors/parts/CreateSpaceAsProcessor.cpp (hunk 8 above)
Comment (bright-starry-sky) [reply src/meta/processors/parts/CreateSpaceAsProcessor.cpp:26]: > OK, Maybe we should not `support if_not_exists` in this sentence?

Yes, I agree not to support if_not_exists. If so, we need to delete this parameter in the relevant place, such as thrift interface.

**Review by bright-starry-sky — APPROVED** [review]
Nice.

**Review by panda-sheep — APPROVED** [review]

**Pull request merged**